tactic/interactive.lean

```diff
@@ -320,5 +320,21 @@ do p' ← to_expr p,
    ; intron (n-2) ; try (solve_by_elim <|> tauto <|> (tactic.intros >> cc)),
    return ()
 
+@[user_attribute]
+meta def funext_attribte : user_attribute :=
+{ name := `extensionality,
+  descr := "lemmas usable by `funext` tactic" }
+
+attribute [extensionality] funext array.ext
+
+meta def ext1 (x : parse ident_ ?) : tactic unit :=
```

Comment:
I think this should be a private definition. Otherwise it will always be slightly confusing if one should use `ext x` or `ext1 x`.

Reply:
I disagree. These two tactics have the same relation as `congr` vs `congr_n`; one does a single step while the other repeats. I think that `ext` should probably repeat through non-intro extensionality rules as well (although it looks like there are no non-intro ext rules in this batch).

Reply:
I don't like this distinction, I write `intro a` and then `intro a b`, but need to fix `intro` to `intros`. I have the feeling the same might happen with `ext1` to `ext`. Why not use `ext _` instead of `ext1`?

Reply:
because not all extensionality rules introduce a bound variable. For example `prod` has an extensionality rule with two hyps and no bound variables, and `ulift` has an extensionality rule with one hyp and no bound variables. I envision `ext` applying these rules eagerly while `ext1` would only apply one of them. Furthermore, `ext1` might need to take more than one BV if the rule calls for it, for example if there are two hyps each with one BV or one hyp with two BVs.

Reply:
@digama0 I like your idea. As it is, I think `ulift` should already work. I am unsure how to handle `prod` because both branches may need different numbers of identifier. The simplest thing might be to just name them in order but I remember it being a pain when I used `inductive` without `case`.

Reply:
... is there a way to have a similar syntax for extensionality? Maybe that would be the right way of handling multiple subgoals.

Reply:
I think both `ext` and `ext1` should just accept a list of names and pull from that list in some deterministic order whenever an intro is done during the course of the tactic. So yes, it would be like the name list in `induction with`. It's not super ergonomic, but it is reliable and handles the simple case well (which is the most likely use anyway).

I don't know if you can tag the subcases so that `case` works with them; it's not obvious what the name would be unless you use the hypothesis name in the type of the theorem.